[#17] DB 연결 및 테스트 코드 작성

canlion opened the pull request:
# 목표
* 엔티티의 레포지터리를 DB 레포지터리로 변경
* 테스트 코드 작성 (testcontainers 활용)
* 네이버 클라우드에 배포

## 엔티티의 레포지터리를 DB 레포지터리로 변경
* 회원, 가게 엔티티의 레포지터리를 HashMap에서 MySQL+JPA(SpringDataJPA) 레포지터리로 변경했습니다.
* 회원, 가게, 음식 엔티티의 속성 중 다른 엔티티와의 관계 표현을 위해 삽입한 식별 번호를 해당 엔티티 객체로 변경했습니다.
   ```java
   @Entity
   public class Store {
      @Id
      @GeneratedValue(strategy = GenerationType.IDENTITY)
      private Long storeId;
      @ManyToOne
      private Member member;  // Long memberId 에서 엔티티 객체로 변경
   ```
  * 컨트롤러에서 DTO를 엔티티로 변경하는 과정에서 연관된 다른 엔티티 객체 필드는 null로 둔 채 서비스 레이어로 전달, 서비스 레이어에서 엔티티 객체를 채우도록 변경했습니다.
    * 컨트롤러에서 해당 엔티티의 레포지터리까지 참조하기보다는 이미 레포지터리를 참조하고있는 서비스 계층에서 몰아서 처리하는게 더 깔끔하다고 생각됩니다.

## 테스트 코드 작성 (testcontainers 활용)
* 회원, 가게 레포지터리에 대한 테스트 코드를 작성했습니다.
* 이전처럼 Testcontainers 컨테이너 생성 과정을 포함하는 슈퍼클래스를 테스트 클래스에서 상속받는 형식으로 구성했는데 컨테이너를 static 필드로 선언하는 경우 첫 자식 클래스의 테스트 이후 두번째 자식 클래스부터는 컨테이너가 동작하지 않아 인스턴스 필드로 변경했습니다.
  * 원인은 아직 못찾았습니다.
* 가게 레포지터리 테스트에서 가게 엔티티가 회원 엔티티와 다대일 관계가 있어 EntityManager를 이용해 회원을 추가하는 방식으로 진행했습니다. TestEntityManager 등을 찾았으나 Testcontainer 슈퍼클래스를 이용하는 현재 구조에서 @DataJpaTest를 적용하는 시도가 실패해 EntityManager를 사용했습니다.

## 배포
* 주중에 크레딧을 신청해 아직 승인이 되지 않았습니다.🙇

# TODO
* 이어서 테스트 코드 작성
* 네이버 클라우드에 배포
* Optional 사용 관련 정리
  * "Optional.orElse"와 "throw"가 별개로 작성된 부분을 "Optional.orElseThrow"로 정리
  * 레포지터리 레이어에서 수행할 작업이 서비스 레이어에 포함되어 Optional이 애매하게 사용되는 부분 정리
    * ex) 가게 삭제: 서비스 레이어에서 가게 존재 여부 확인, 레포지터리에서 가게 조회시에 "Optional.get" 사용하여 삭제된 가게 정보 반환 -> 레포지터리에서 "Optional.orElseThrow"로 존재 여부 확인까지 수행하도록 정리
* 서비스간의 순환참조가 일어나지 않도록 참조 관계 다시 고려해보기

canlion committed:
feat: 테스트용 MySQL 컨테이너 설정 클래스 추가

- Testcontainers MySQL 컨테이너 환경 구성
- 레포지터리 테스트 클래스에서 상속하여 사용

canlion committed:
fix: 오타 수정

canlion committed:
feat: SpringDataJPA/MySQL 회원 레포지터리 적용

- 회원 엔티티 관리 방식을 HashMap에서 SpringDataJPA/MySQL로 변경

canlion committed:
feat: 가게 엔티티의 회원 번호 속성을 회원 객체로 변경

- 연관관계 표현을 위해 `Long memberId`를 `Member member`로
  변경
  - 가게 엔티티의 `member` 속성은 서비스 계층에서 회원 조회하여 할당

canlion committed:
feat: SpringDataJPA/MySQL 가게 레포지터리 적용

- 가게 엔티티 관리 방식을 HashMap에서 SpringDataJPA/MySQL로 변경

canlion committed:
feat: 음식 엔티티의 가게 번호 속성을 가게 객체로 변경

- 연관관계 표현을 위해 `Long storeId`를 `Store store`로 변경
  - `Store store` 속성은 서비스 계층에서 할당

canlion committed:
test: 회원 레포지터리 테스트 코드 추가

canlion committed:
fix: MySQL 컨테이너 객체를 인스턴스 필드로 변경

- 컨테이너를 static으로 선언하는 경우 여러 테스트 클래스에서 상속하여
  사용할 때 컨테이너가 동작하지 않음.
  - 원인 파악 후 재시도 예정

canlion committed:
test: 가게 레포지터리 테스트 코드 추가

canlion committed:
fix: 가게 서비스 레이어 @Transactional 적용

f-lab-lyan commented on api/src/main/java/burgermap/repository/MySqlStoreRepository.java:
```diff
@@ -0,0 +1,45 @@
+    @Override
+    public Optional<Store> updateStore(Long storeId, Store newStore) {
+        newStore.setStoreId(storeId);
+        repository.save(newStore);
+        return Optional.of(newStore);
```
```suggestion
        return repository.save(newStore);
```

canlion committed:
style: 불필요한 주석 제거

canlion committed:
chore: 테스트 데이터 초기화 방식을 sql 스크립트에서 코드 기반으로 변경

- 런타임을 위한 init.sql은 변경되기 쉬우므로 직접 코드에서 데이터 삽입 수행

canlion committed:
test: 재료 레포지터리 테스트 코드 추가

canlion committed:
test: 재료 레포지터리 테스트 케이스 추가

- 케이스 "존재하지 않는 재료 번호로 재료 조회" 추가

canlion committed:
test: 메뉴카테고리 레포지터리 테스트 코드 추가

canlion committed:
test: 회원 서비스 테스트 코드 추가

canlion committed:
test: 회원 서비스의 누락된 테스트 추가

- isMemberTypeOwner 메서드에 대한 테스트 누락되어 추가

f-lab-lyan commented on api/src/test/java/burgermap/service/MemberServiceTest.java:
```diff
@@ -0,0 +1,257 @@
+        });
+        Member result = memberService.changeNickname(memberId, newNickname);
+
+        assertThat(result.getNickname()).isEqualTo(newNickname);
```
verify(memberService.updateNickName(any(), any()), times(1));

canlion committed:
[#20] 레포지터리 레이어에 포함된 엔티티 업데이트 메서드를 서비스 레이어로 이동 (#21)

* refactor: 회원 정보 변경 메서드를 서비스 레이어로 이동

- 엔티티 수정은 새로운 값의 검증 등 비즈니스 로직 결과를 반영하는
  작업으로 서비스 레이어에서 수행하고, 레포지터리는 순수하게 DB와의
상호작용만을 담당하도록 함.

* refactor: 가게 정보 변경 메서드를 서비스 레이어로 이동

- 엔티티 수정은 비즈니스 로직의 결과를 반영하는 작업으로 서비스
  레이어에서 수행, 레포지터리는 DB와의 상호작용만을 담당하도록 하여
레이어의 역할을 명확히 함.

* test: 제거된 메서드의 테스트 제거

* 엔티티 수정 관련 메서드를 서비스로 옮기면서 레포지터리 테스트에서
  테스트 제거
  - 가게 엔티티 수정 메서드 테스트 케이스 제거
  - 회원 엔티티 수정 메서드 테스트 케이스 제거

canlion committed:
fix: 로직 변경을 테스트에 반영

- 서비스가 참조하던 레포지터리의 메서드를 제거하며 서비스 로직 수정,
  이를 테스트에 반영함
  - MemberService의 회원 정보 수정 메서드 로직 변경
    - changePassword
    - changeEmail
    - changeNickname

canlion committed:
[#19] Optional의 사용 개선 및 로직 정리

Squashed commit of the following:

commit 03463866d9731a7beb7bf2f4ebd1186e71de93ad
Author: canlion <canlion@naver.com>
Date:   Sat Oct 19 19:44:52 2024 +0900

    refactor: 등록되지 않은 재료 필터링 로직 단순화

commit dca5d58fedea6fffafb23c09b386791e950db17e
Author: canlion <canlion@naver.com>
Date:   Sat Oct 19 17:44:27 2024 +0900

    style: 코드 스타일 통일을 위해 줄바꿈 적용

commit e7c768958b45a2f9150ea7efd81d918d59fc88f9
Author: canlion <canlion@naver.com>
Date:   Sat Oct 19 17:25:11 2024 +0900

    refactor: 불필요한 Optional 사용 제거

    - 가게 삭제에 앞서 가게의 존재를 검증하므로 Optional 사용이 불필요하여
      제거, 가게 존재 검증시 조회한 가게 정보를 반환하도록 수정

commit 6c187ec2972068508cccbf111e443f59b3d9ab18
Author: canlion <canlion@naver.com>
Date:   Sat Oct 19 17:22:54 2024 +0900

    refactor: 불필요한 조건문 제거

commit 4288c2854c4480e82ad6ffa991dde9214705a805
Author: canlion <canlion@naver.com>
Date:   Sat Oct 19 16:31:45 2024 +0900

    refactor: 회원 레포지터리에서 Optional.get 제거

    - 서비스에서 조회 결과의 존재 검증이 선행되어 Optional.get을 사용했으나
      서비스와 레포지터리의 결합이 강해지고 서비스에 종속됨.
    - Optional을 그대로 반환하도록 수정

commit aea847a185fdf5503b85045e67d373c692d77ca5
Author: canlion <canlion@naver.com>
Date:   Sat Oct 19 15:39:21 2024 +0900

    refactor: 로그인 메서드의 로그, Optional 사용 개선

    - 로그인 실패시 패스워드 오류와 로그인 아이디 오류를 나누어 로깅
    - Optional.get을 Optional.isPresent 하위에서 수행해 코드 흐름을 알기
      쉽도록 변경

commit c9758d467df7ed173111ffc6867956b8cb7246b4
Author: canlion <canlion@naver.com>
Date:   Thu Oct 17 09:50:19 2024 +0900

    refactor: 회원 레포지터리에서의 NoSuchElementException 발생 제거

    - 회원 정보의 유무에 따른 대응을 서비스레이어에서 수행해 레이어간 책임을
      명확히 분리

commit f61be8ee0c6ffb48e10593d75aaf5ab7040254e5
Author: canlion <canlion@naver.com>
Date:   Thu Oct 17 09:47:43 2024 +0900

    refactor: 회원 조회 결과가 없는 경우 예외 발생시킴

    - 서비스 레이어에서 MemberRepository.findByMemberId 결과가
      Optional.empty인 경우 MemberNotExistException 발생시킴
      - 인터셉터에서 로그인 검증을 선행하나 회원 서비스 코드만으로는 알 수
        없음.
      - 인터셉터의 동작을 고려하지 않고 서비스 레이어와 다른 요소들간의
        결합도를 낮춤.

commit 94061836513fe360fc139d4316eb3f403c62baca
Author: canlion <canlion@naver.com>
Date:   Thu Oct 17 09:29:13 2024 +0900

    feat: MemberNotExistException & 회원 예외 advice 추가

    - MemberNotExistException: 주어진 회원 식별 번호의 회원 조회 결과가 없을 경우 throw
    - MemberExceptionAdvice: 회원 관련 예외 처리

canlion commented:
서비스 로직에 `Optional`을 일관성 없이 사용한 부분이 있었는데 테스트 코드 작성할 때 걸림돌이 되어 `Optional`을 사용한 부분들을 먼저 개선했습니다. 레포지터리가 서비스에, 서비스가 인터셉터와 컨트롤러에 의존하지 않도록 코드를 변경했습니다.
* 예시 - `StoreService.addStore`
  * 가게 등록을 요청한 회원이 `Owner` 타입인지 검증할 때에는 주어진 회원 번호를 가진 회원이 존재한다고 가정하여 존재하지 않는 회원 번호를 조회하여 발생하는 예외를 가정하지 않음. (인터셉터에서 로그인 검증이 선행된다는 점에 의존)
  * 가게 등록을 위해 주어진 정보에 회원 엔티티를 주입할 때에는 주어진 회원 번호가 반드시 존재한다는 가정이 없어 존재하지 않는 회원번호를 조회했을 때 예외를 발생시킴.
  * 코드에서 조회하려는 엔티티가 무조건 존재한다는 가정와 그렇지 않다는 가정이 혼재하여 테스트 작성시에 회원의 타입 검증에 대해서는 예외를 가정하지 않고, 회원 조회시에는 예외를 가정하면서 테스트 코드도 일관성이 없어짐.

f-lab-lyan commented on api/src/main/java/burgermap/service/MemberService.java:
```diff
@@ -50,54 +51,64 @@ public boolean checkNicknameDuplication(String nickname) {
-        if (member.isEmpty() || !member.get().getPassword().equals(password)) {
-            log.debug("login failed");
-            return null;
+        if (member.isPresent()) {
```
Optional.map(....).orElseGet(null);

f-lab-lyan replied:
member
   .filter(member => member.getPassword().equals(password))
   .orElseGet(null);
